src/app/icd/client/RefreshKeySender.cpp

```diff
@@ -0,0 +1,144 @@
+    mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, CHIP_NO_ERROR);
+}
+void RefreshKeySender::Callback::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+    mpRefreshKeySender->GetCheckInDelegate()->OnKeyRefreshDone(mpRefreshKeySender, mError);
+}
+
+void RefreshKeySender::Callback::OnDone(chip::app::CommandSender * apCommandSender) {}
+
+CHIP_ERROR RefreshKeySender::RegisterClientWithNewKey(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    using namespace Clusters::IcdManagement;
+    EndpointId endpointId = 0;
+    GroupId groupId       = 0;
+
+    mRegisterCommandSender.Emplace(&mCommandSenderDelegate, &exchangeMgr);
+
+    mCommandSenderDelegate.SetRefreshKeySender(this);
+
+    Commands::RegisterClient::Type registerClientCommand;
+    registerClientCommand.checkInNodeID    = mICDClientInfo.peer_node.GetNodeId();
+    registerClientCommand.monitoredSubject = mICDClientInfo.monitored_subject;
+    registerClientCommand.key              = GetRefreshKeyBuffer().Span();
+    auto commandPathParams =
+        CommandPathParams(endpointId, groupId, Id, Commands::RegisterClient::Id, (CommandPathFlags::kEndpointIdValid));
+
+    ReturnErrorOnFailure(mRegisterCommandSender.Value().AddRequestData(commandPathParams, registerClientCommand));
+    ReturnErrorOnFailure(mRegisterCommandSender.Value().SendCommandRequest(sessionHandle));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RefreshKeySender::EstablishSessionToPeer()
+{
+    ChipLogProgress(ICD, "Trying to establish a CASE session for re-registering an ICD client");
+    auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+    VerifyOrReturnError(caseSessionManager != nullptr, CHIP_ERROR_INVALID_CASE_PARAMETER);
+    caseSessionManager->FindOrEstablishSession(mICDClientInfo.peer_node, &mOnConnectedCallback, &mOnConnectionFailureCallback);
+    return CHIP_NO_ERROR;
+}
+
+void RefreshKeySender::HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                             const SessionHandle & sessionHandle)
+{
+    RefreshKeySender * const _this = static_cast<RefreshKeySender *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    CHIP_ERROR err = _this->RegisterClientWithNewKey(exchangeMgr, sessionHandle);
+    if (CHIP_NO_ERROR != err)
+    {
+        ChipLogError(ICD, "Failed to send register client command");
+        _this->mpCheckInDelegate->OnKeyRefreshDone(_this, err);
+    }
+}
+
+void RefreshKeySender::HandleDeviceConnectionFailure(void * context, const ScopedNodeId & peerId, CHIP_ERROR err)
+{
+    RefreshKeySender * const _this = static_cast<RefreshKeySender *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    ChipLogError(ICD, "Failed to establish CASE for re-registration with error '%" CHIP_ERROR_FORMAT "'", err.Format());
```

Comment:
establish CASE for re-registration ==> establish CASE for ICD key re-registration?